abstract/encoding.go

```diff
@@ -170,13 +193,31 @@ func (de *decoder) value(v reflect.Value, depth int) error {
 			}
 		}
 
-	case reflect.Slice, reflect.Array:
+	case reflect.Slice:
+		if v.IsNil() {
+			panic("slices must be initialized to correct length before decoding")
+		}
+		fallthrough
+	case reflect.Array:
 		l := v.Len()
 		for i := 0; i < l; i++ {
 			if err = de.value(v.Index(i), depth+1); err != nil {
 				return err
 			}
 		}
+
+	case reflect.Int:
```

Comment:
I think I'm still agreeing with Nicolas that from a language/programmability perspective it's just pretty inconvenient for the common 'int' to be not supported at all by the encoding.  

Making it a selectable, perhaps non-default option might be a reasonable alternative.  (I would pick something more specific than "unsafe", since there are many possible things that can be done that are "unsafe" and many are probably more unsafe than int conversions. :) )  Perhaps "AllowInt" or something like that.  Maybe even an "IntSize" enum-type with options "NoInt" (disallowed, default), "Int32", or "Int64".  I haven't yet convinced myself this is worth the extra complexity though.

Yet another thought is to support int encoding but make it go the other way, defining the wire encoding of an int as being always 32 bits.  This would ensure that a protocol that gets adequately tested on 64-bit platform will not break because of smaller ints if later run on a 32-bit platform.  Defining int as 32-bit would also be consistent with Java/Scala etc, where int is always 32 bits.